src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/ProjectSystem/ProjectConfigurationStateManager.cs

```diff
@@ -0,0 +1,181 @@
+// Licensed under the MIT license. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Razor.ProjectSystem;
+using Microsoft.AspNetCore.Razor.Serialization;
+using Microsoft.AspNetCore.Razor.Utilities;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.CodeAnalysis.Razor.Logging;
+using Microsoft.CodeAnalysis.Razor.ProjectSystem;
+using Microsoft.CodeAnalysis.Razor.Utilities;
+using Microsoft.VisualStudio.Threading;
+
+namespace Microsoft.AspNetCore.Razor.LanguageServer.ProjectSystem;
+
+/// <summary>
+/// Updates project system service with new project info.
+/// </summary>
+/// <remarks>
+/// Used to figure out if the project system data is being added, updated or deleted
+/// and acts accordingly to update project system service.
+/// </remarks>
+internal partial class ProjectConfigurationStateManager : IDisposable
+{
+    private readonly IRazorProjectService _projectService;
+    private readonly IProjectSnapshotManager _projectManager;
+    private readonly ILogger _logger;
+
+    private readonly AsyncBatchingWorkQueue<(ProjectKey ProjectKey, RazorProjectInfo? ProjectInfo)> _workQueue;
+    private readonly CancellationTokenSource _disposalTokenSource;
+    private static readonly TimeSpan s_enqueueDelay = TimeSpan.FromMilliseconds(250);
+
+    public ProjectConfigurationStateManager(
+        IRazorProjectService projectService,
+        ILoggerFactory loggerFactory,
+        IProjectSnapshotManager projectManager)
+        : this(projectService,
+               loggerFactory,
+               projectManager,
+               s_enqueueDelay)
+    { }
+
+    // Provided for tests to specify enqueue delay
+    public ProjectConfigurationStateManager(
+        IRazorProjectService projectService,
+        ILoggerFactory loggerFactory,
+        IProjectSnapshotManager projectManager,
+        TimeSpan enqueueDelay)
+    {
+        _projectService = projectService;
+        _projectManager = projectManager;
+        _logger = loggerFactory.GetOrCreateLogger<ProjectConfigurationStateManager>();
+
+        _disposalTokenSource = new ();
+        _workQueue = new(
+            enqueueDelay,
+            ProcessBatchAsync,
+            _disposalTokenSource.Token);
+    }
+
+    public void Dispose()
+    {
+        _disposalTokenSource.Cancel();
+        _disposalTokenSource.Dispose();
+    }
+
+    private ValueTask ProcessBatchAsync(ImmutableArray<(ProjectKey ProjectKey, RazorProjectInfo? ProjectInfo)> workItems, CancellationToken cancellationToken)
+    {
+        foreach (var workItem in workItems.GetMostRecentUniqueItems(Comparer.Instance))
+        {
+            if (cancellationToken.IsCancellationRequested)
+            {
+                return default;
+            }
+
+            UpdateProject(workItem.ProjectKey, workItem.ProjectInfo, cancellationToken);
```

Comment:
Consider making `UpdateProject` and `ResetProject` be properly async and task returning, so this call here is the only one with `.Forget()`. That would help readability, as its reasonably obvious why we don't need to wait in this location, but inside `UpdateProject` I have to track my way back out to try to discover why the tasks aren't awaited. Perhaps even a comment above this line :)

Reply:
Agreed, done.